Gems/GradientSignal/Code/Source/Editor/EditorGradientBakerComponent.cpp

```diff
@@ -125,60 +129,80 @@ namespace GradientSignal
                 // imageBoundsY)
                 AZ::Vector3 uvw(static_cast<float>(x), static_cast<float>((imageResolutionY - 1) - y), 0.0f);
 
-                GradientSampleParams sampleParams;
-                sampleParams.m_position = inputBoundsStart + (uvw * pixelToBoundsScale) + scaledTexelOffset;
+                AZ::Vector3 position = inputBoundsStart + (uvw * pixelToBoundsScale) + scaledTexelOffset;
 
                 bool inBounds = true;
                 LmbrCentral::ShapeComponentRequestsBus::EventResult(
-                    inBounds, m_boundsEntityId, &LmbrCentral::ShapeComponentRequestsBus::Events::IsPointInside, sampleParams.m_position);
+                    inBounds, m_boundsEntityId, &LmbrCentral::ShapeComponentRequestsBus::Events::IsPointInside, position);
 
-                float sample = inBounds ? m_configuration.m_gradientSampler.GetValue(sampleParams) : 0.0f;
-
-                // Write out the sample value for the pixel based on output format
-                int index = ((y * imageResolutionX) + x) * channels;
-                switch (m_configuration.m_outputFormat)
-                {
-                case OutputFormat::R8:
+                if (!inBounds)
                 {
-                    AZ::u8 value = static_cast<AZ::u8>(sample * std::numeric_limits<AZ::u8>::max());
-                    pixels[index] = value; // R
-
-                    if (channels == 4)
-                    {
-                        pixels[index + 1] = value; // G
-                        pixels[index + 2] = value; // B
-                        pixels[index + 3] = std::numeric_limits<AZ::u8>::max(); // A
-                    }
-                    break;
+                    continue;
                 }
-                case OutputFormat::R16:
+
+                // Keep track of this input position + the x,y indices
+                inputPositions.push_back(position);
+                indices.push_back(AZStd::make_pair(x, y));
+            }
+        }
+
+        // Retrieve all the gradient values for the input positions
+        const size_t numPositions = inputPositions.size();
+        AZStd::vector<float> outputValues(numPositions);
+        m_configuration.m_gradientSampler.GetValues(inputPositions, outputValues);
+
+        // Write out all the gradient values to our output image
+        for (int i = 0; !m_shouldCancel && (i < numPositions); ++i)
+        {
+            const float& sample = outputValues[i];
+            const auto& indicesPair = indices[i];
+            const auto& x = indicesPair.first;
+            const auto& y = indicesPair.second;
+
+            // Write out the sample value for the pixel based on output format
+            int index = ((y * imageResolutionX) + x) * channels;
+            switch (m_configuration.m_outputFormat)
```

Comment:
nit: one idea to clean this up a bit and maaaaybe make it faster would be to put each output format in a separate function (or lambda), determine which function needs to be called before the for loop and bind it to a callback that you call inside the loop. That way there'd be no switch / or branching inside the loop and the unpacking code could possibly be reused later.

Reading through this a bit more, you really would only need on templated function that templates based on the output type and the max value.

Reply:
FWIW, I tried something similar to that in GradientTransform I think, and the switch actually turned out to be faster than the function pointer call. So just be wary that YMMV with this one. :) At a minimum, moving the code into separate static functions for reusability didn't hurt the performance in my case.